Experiment definition framework

k88hudson opened the pull request:
This patch adds a experiment definition framework in the form of a json file.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/activity-stream/886)
<!-- Reviewable:end -->

k88hudson commented on test/test-ExperimentProvider.js:
```diff
@@ -0,0 +1,105 @@
+  }, "clientID '012j' should result in variant being picked");
+};
+
+// exports["test ExperimentProvider.validateExperiment"] = assert => {
```
`assert.throws` seems to intermittently fuck up the tests, I get this output sometimes:

```
console.error: activity-streams:
  warnings and/or errors were logged.

26 of 27 tests passed.
There were test failures...

The following tests failed:


```

i.e., it causes a failure, but no tests actually failed?

k88hudson commented:
@ncloudioj want to take a look? I wrote some documentation in https://github.com/k88hudson/activity-stream/blob/fa451804114b2dca71ea823462aa3ea006320b82/experiments_how_to.md

k88hudson commented:
@jaredkerim suggested that maybe we just automatically add the cohort/experiment ID to every ping instead of requiring a specific code path for it/a whitelist. what do you think @ncloudioj?

ncloudioj commented on lib/ExperimentProvider.js:
```diff
@@ -0,0 +1,59 @@
+        return;
+      }
+      const {id, variant, control} = experiment;
+      const inExperiment = this._rng() <= variant.threshold;
```
A single threshold is fine for the single experiment case, but we need a better experiment user selection mechanism to support those multi-experiment scenarios.

For instance, say we have three independent experiments - e1, e2, e3 running at the same time, each of them is targeting 10% of the addon users, respectively. Ideally, the whole user base should be divided into four groups as follows,
* control group, if 0 <= `rng` < 0.7
* e1 group, if 0.7 <= `rng` < 0.8
* e2 group, if 0.8 <= `rng` < 0.9
* e3 group otherwise

ncloudioj commented on lib/ExperimentProvider.js:
```diff
@@ -0,0 +1,59 @@
+    if (experiment.variant && typeof experiment.variant.threshold !== "number") {
+      errors.push("missing variant.threshold, which should be a number");
+    }
+    if (experiment.variant && experiment.variant.threshold > 1) {
```
also, threshold > 0

ncloudioj commented:
LGTM except a few nits. This framework largely simplifies the work of experimenting. Awesome work!!!

k88hudson commented:
OK I added:
* automatic `experiment_id` added to ping
* tests/validator that will throw if the total control cohort is less than 20%
* moved `id` to variant, so if we want to add multiple variants in the future, this is easy to support

ncloudioj commented:
r+ :shipit:
